apps/services/bff/src/app/modules/auth/token-refresh.service.ts

```diff
@@ -0,0 +1,213 @@
+    return new Promise((resolve, reject) => {
+      const timeoutId = setTimeout(() => {
+        clearInterval(pollInterval)
+        reject(
+          new Error(
+            `Polling timed out for token refresh completion for session ${sid}`,
+          ),
+        )
+      }, TokenRefreshService.MAX_POLL_TIME)
+
+      const pollInterval = setInterval(async () => {
+        try {
+          const refreshTokenInProgress = await this.cacheService.get<boolean>(
+            this.createRefreshTokenKey(sid),
+            false,
+          )
+
+          if (!refreshTokenInProgress) {
+            clearInterval(pollInterval)
+            clearTimeout(timeoutId)
+            resolve()
+          }
+        } catch (error) {
+          clearInterval(pollInterval)
+          clearTimeout(timeoutId)
+          reject(new Error(`Error polling for refresh completion: ${error}`))
+        }
+      }, TokenRefreshService.POLL_INTERVAL)
+    })
+  }
+
+  /**
+   * Handles the complete token refresh process with concurrent request protection
+   * This method:
+   *
+   * 1. Checks if a refresh is already in progress
+   * 2. If yes, waits for it to complete
+   * 3. If no, initiates a new refresh
+   * 4. Updates the cache with new token data
+   * 5. Cleans up tracking flags
+   *
+   * @param params.sid - Session ID
+   * @param params.encryptedRefreshToken - Encrypted refresh token to use for getting new tokens
+   *
+   * @returns Promise resolving to updated token response data
+   * @throws Forwards any errors from the refresh process after logging
+   */
+  public async refreshToken({
+    sid,
+    encryptedRefreshToken,
+  }: {
+    sid: string
+    encryptedRefreshToken: string
+  }): Promise<CachedTokenResponse> {
+    const refreshTokenKey = this.createRefreshTokenKey(sid)
+    const tokenResponseKey = this.createTokenResponseKey(sid)
+
+    try {
+      // Check if refresh is already in progress
+      const refreshTokenInProgress = await this.cacheService.get<boolean>(
+        refreshTokenKey,
+        false,
+      )
+
+      if (refreshTokenInProgress) {
+        try {
+          // Wait for the ongoing refresh to complete
+          await this.pollForRefreshCompletion(sid)
+
+          // Get the updated token response from cache
+          const tokenResponse =
+            await this.cacheService.get<CachedTokenResponse>(
+              tokenResponseKey,
+              false,
+            )
+
+          if (!tokenResponse) {
+            throw new Error(
+              `Refresh token polling completed but token data not found for sid: ${sid}`,
+            )
+          }
+
+          return tokenResponse
+        } catch (error) {
+          this.logger.error(error)
+
+          // If polling times out, then retry the refresh
+          const updatedTokenResponse = await this.executeTokenRefresh({
+            refreshTokenKey,
+            encryptedRefreshToken,
+          })
+
+          return updatedTokenResponse
```

Comment:
~Unnecessary code duplication. Enough to log a warning and continue below to refresh the token.